dotnet/src/InternalUtilities/src/Diagnostics/ModelDiagnostics.cs

```diff
@@ -35,6 +36,26 @@ internal static class ModelDiagnostics
     private static readonly bool s_enableDiagnostics = AppContextSwitchHelper.GetConfigValue(EnableDiagnosticsSwitch, EnableDiagnosticsEnvVar);
     private static readonly bool s_enableSensitiveEvents = AppContextSwitchHelper.GetConfigValue(EnableSensitiveEventsSwitch, EnableSensitiveEventsEnvVar);
 
+    /// <summary>
+    /// Stores streaming text content for activities of streaming completions.
+    /// </summary>
+    private static readonly Dictionary<string, List<StreamingKernelContent>> s_streamingContents = [];
+
+    static ModelDiagnostics()
+    {
+        s_activityListener.ShouldListenTo = activitySource => activitySource.Name == s_namespace;
+        s_activityListener.ActivityStopped = activity =>
+        {
+            // Called when an activity is stopped. Clean up the streaming content in case `EndStreaming` is not called.
+            // This action needs to be idempotent as the event may be fired multiple times.
+            if (activity.Id is not null)
+            {
+                s_streamingContents.Remove(activity.Id);
+            }
+        };
+        ActivitySource.AddActivityListener(s_activityListener);
```

Comment:
Using a listener here seems like unnecessary overhead and complication. Can't the client just have a StringBuilder that it appends to and then passes that off at the end?

Reply:
Note as well that Dictionary is not thread-safe. Concurrency here could end up corrupting s_streamingContents.

Reply:
I think it could be beneficial to us if we do handle merging the chunks and handling the formatting in the diagnostics class as opposed to in each client. For example, we don't have to merge the chunks if model diagnostics is not enabled. And we perform the checks in a single place. 

In addition, we will need other metadata as well, not just the texts.

Reply:
You don't have to do it in each client. You just need to have the state stored in the client and pass it in to a method that does the work.

Reply:
(Regardless, what's currently there can't be merged; it's not safe.)

Reply:
> Note as well that Dictionary is not thread-safe. Concurrency here could end up corrupting s_streamingContents.

Since the keys of the dictionary are the activity IDs, which are unique across all processes, would concurrency be a concern here?

Reply:
Dictionary is not thread safe. You can't safely read/write to it from multiple threads concurrently, regardless of which keys they're using.

Reply:
(It's also a possible leak. If, for example, an async enumerator is just dropped rather than being pushed to a final state, I expect this is leaking data into a Dictionary and it'll never be cleaned up.)

Reply:
Would `ConcurrentDictionary` make it thread safe?

Reply:
It'd avoid corrupting the dictionary for concurrent mutations to the dictionary itself. I suspect you could still end up in a situation where multiple threads are trying to modify the list for the same activity, e.g. if some code called an async method but didn't await it and then proceeded to do something else, both forks of execution will share the same async locals and activity.

This is still a possible leak, though.

And is still unnecessary overhead.

It's clever, but I don't think it's the right approach.